.dotnet/src/Custom/Batch/BatchOperation.Protocol.cs

```diff
@@ -0,0 +1,287 @@
+    /// <returns> The response returned from the service. </returns>
+    public virtual ClientResult GetBatch(string batchId, RequestOptions options)
+    {
+        Argument.AssertNotNullOrEmpty(batchId, nameof(batchId));
+
+        using PipelineMessage message = CreateRetrieveBatchRequest(batchId, options);
+        return ClientResult.FromResponse(_pipeline.ProcessMessage(message, options));
+    }
+
+    /// <summary>
+    /// [Protocol Method] Cancels an in-progress batch.
+    /// </summary>
+    /// <param name="batchId"> The ID of the batch to cancel. </param>
+    /// <param name="options"> The request options, which can override default behaviors of the client pipeline on a per-call basis. </param>
+    /// <exception cref="ArgumentNullException"> <paramref name="batchId"/> is null. </exception>
+    /// <exception cref="ArgumentException"> <paramref name="batchId"/> is an empty string, and was expected to be non-empty. </exception>
+    /// <exception cref="ClientResultException"> Service returned a non-success status code. </exception>
+    /// <returns> The response returned from the service. </returns>
+    public virtual async Task<ClientResult> CancelBatchAsync(string batchId, RequestOptions options)
+    {
+        Argument.AssertNotNullOrEmpty(batchId, nameof(batchId));
+
+        using PipelineMessage message = CreateCancelBatchRequest(batchId, options);
+        return ClientResult.FromResponse(await _pipeline.ProcessMessageAsync(message, options).ConfigureAwait(false));
+    }
+
+    /// <summary>
+    /// [Protocol Method] Cancels an in-progress batch.
+    /// </summary>
+    /// <param name="batchId"> The ID of the batch to cancel. </param>
+    /// <param name="options"> The request options, which can override default behaviors of the client pipeline on a per-call basis. </param>
+    /// <exception cref="ArgumentNullException"> <paramref name="batchId"/> is null. </exception>
+    /// <exception cref="ArgumentException"> <paramref name="batchId"/> is an empty string, and was expected to be non-empty. </exception>
+    /// <exception cref="ClientResultException"> Service returned a non-success status code. </exception>
+    /// <returns> The response returned from the service. </returns>
+    public virtual ClientResult CancelBatch(string batchId, RequestOptions options)
+    {
+        Argument.AssertNotNullOrEmpty(batchId, nameof(batchId));
+
+        using PipelineMessage message = CreateCancelBatchRequest(batchId, options);
+        return ClientResult.FromResponse(_pipeline.ProcessMessage(message, options));
+    }
+
+    internal PipelineMessage CreateRetrieveBatchRequest(string batchId, RequestOptions options)
+    {
+        var message = _pipeline.CreateMessage();
+        message.ResponseClassifier = PipelineMessageClassifier200;
+        var request = message.Request;
+        request.Method = "GET";
+        var uri = new ClientUriBuilder();
+        uri.Reset(_endpoint);
```

Comment:
can we add a ctor parameter that takes the Uri?

Reply:
Sure -- this is also a type emitted by the generator, but that should be possible.  I can file an issue in the generator repo to do this.

Reply:
Filed https://github.com/Azure/autorest.csharp/issues/4968 to track.